osu.Game/Screens/Play/HUD/MultiplayerGameplayLeaderboard.cs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using JetBrains.Annotations;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Game.Configuration;
+using osu.Game.Database;
+using osu.Game.Online.Spectator;
+using osu.Game.Rulesets.Scoring;
+
+namespace osu.Game.Screens.Play.HUD
+{
+    public class MultiplayerGameplayLeaderboard : GameplayLeaderboard
+    {
+        private readonly ScoreProcessor scoreProcessor;
+
+        private readonly int[] userIds;
+
+        private readonly Dictionary<int, TrackedUserData> userScores = new Dictionary<int, TrackedUserData>();
+
+        /// <summary>
+        /// Construct a new leaderboard.
+        /// </summary>
+        /// <param name="scoreProcessor">A score processor instance to handle score calculation for scores of users in the match.</param>
+        /// <param name="userIds">IDs of all users in this match.</param>
+        public MultiplayerGameplayLeaderboard(ScoreProcessor scoreProcessor, int[] userIds)
+        {
+            // todo: this will eventually need to be created per user to support different mod combinations.
+            this.scoreProcessor = scoreProcessor;
+
+            // todo: this will likely be passed in as User instances.
+            this.userIds = userIds;
+        }
+
+        [Resolved]
+        private SpectatorStreamingClient streamingClient { get; set; }
+
+        [Resolved]
+        private UserLookupCache userLookupCache { get; set; }
+
+        private Bindable<ScoringMode> scoringMode;
+
+        [BackgroundDependencyLoader]
+        private void load(OsuConfigManager config)
+        {
+            streamingClient.OnNewFrames += handleIncomingFrames;
+
+            foreach (var user in userIds)
+            {
+                streamingClient.WatchUser(user);
+
+                // probably won't be required in the final implementation.
+                var resolvedUser = userLookupCache.GetUserAsync(user).Result;
```

Comment:
Slightly alarmed by the `.Result`. Not saying this will necessarily deadlock (and I haven't thought very hard if it can), but may be wise to at least stick a `[LongRunningLoad]` on this for the time being if we're going with this simplification for now?

Reply:
Only this way with the comment above attached to it. As discussed with smoogi, the users will likely be provided in the ctor rather than raw IDs, at which point this code will be removed.